src/main/java/net/sf/jabref/remote/DBMSProcessor.java

```diff
@@ -0,0 +1,473 @@
+        }
+
+        query.append(" WHERE ");
+        query.append(escape(ENTRY_REMOTE_ID));
+        query.append(" = ");
+        query.append(bibEntry.getRemoteId());
+        dbmsHelper.executeUpdate(query.toString());
+    }
+
+    /**
+     * Removes the remote existing bibEntry
+     * @param bibEntry {@link BibEntry} to be deleted
+     */
+    public void removeEntry(BibEntry bibEntry) {
+        String query = "DELETE FROM " + escape(ENTRY) + " WHERE " + escape(ENTRY_REMOTE_ID) + " = "
+                + bibEntry.getRemoteId();
+        dbmsHelper.executeUpdate(query);
+        normalizeEntryTable();
+    }
+
+    /**
+     *  Prepares the database table for a new {@link BibEntry}.
+     *  Learning table structure: Columns which are not available are going to be created.
+     *
+     *  @param bibEntry Entry which pretends missing columns which should be created.
+     *
+     */
+    public void prepareEntryTableStructure(BibEntry bibEntry) {
+        Set<String> fieldNames = dbmsHelper.allToUpperCase(bibEntry.getFieldNames());
+        fieldNames.removeAll(dbmsHelper.allToUpperCase(dbmsHelper.getColumnNames(escape(ENTRY))));
+
+        String columnType = dbmsType == DBMSType.ORACLE ? " CLOB NULL" : " TEXT NULL DEFAULT NULL";
+
+        for (String fieldName : fieldNames) {
+            dbmsHelper.executeUpdate("ALTER TABLE " + escape(ENTRY) + " ADD " + escape(fieldName) + columnType);
+        }
+    }
+
+    /**
+     *  Deletes all unused columns where every entry has a value NULL.
+     */
+    public void normalizeEntryTable() {
+        ArrayList<String> columnsToRemove = new ArrayList<>();
+
+        columnsToRemove.addAll(dbmsHelper.allToUpperCase(dbmsHelper.getColumnNames(escape(ENTRY))));
+        columnsToRemove.remove(ENTRY_REMOTE_ID); // essential column
+        columnsToRemove.remove(ENTRY_ENTRYTYPE); // essential column
+
+        try (ResultSet resultSet = selectFromEntryTable()) {
+            while (resultSet.next()) {
+                for (int i = 0; i < columnsToRemove.size(); i++) {
+                    if (resultSet.getObject(columnsToRemove.get(i)) != null) {
+                        columnsToRemove.remove(i);
+                        i--; // due to index shift
+                    }
+                }
+            }
+        } catch (SQLException e) {
+            LOGGER.error("SQL Error: ", e);
+        }
+
+        dropColumns(columnsToRemove);
+    }
+
+    /**
+     * Converts all remotely present bib entries to the List of real {@link BibEntry} objects and retrieves them.
+     */
+    public List<BibEntry> getRemoteEntries() {
+        return getRemoteEntries(0);
+    }
+
+    /**
+     * @param remoteId Entry ID
+     * @return instance of {@link BibEntry}
+     */
+    public BibEntry getRemoteEntry(int remoteId) {
+        List<BibEntry> entries = getRemoteEntries(remoteId);
+        if (entries.size() > 0) {
+            return entries.get(0);
```

Comment:
Because `remoteId` is unique within the remote database. Now it returns only `Optional.empty()` if the entry was not found.